### PR TITLE
Flat CSV Exporter

### PR DESCRIPTION
**Pull request by jercoh**

### CODE REVIEW

File: solvebio/exporters.py [hunk 1]
```diff
@@ -200,6 +215,119 @@ def write(self, filename):
             f.close()
 
 
+class FlatCSVExporter(object):
```
Comment (davecap) [inline solvebio/exporters.py]: is it possible to subclass `CSVExporter` and use `__init__` and `write()` from the parent class?

Comment (jercoh) [inline solvebio/exporters.py]: Sure I'll fix this

File: solvebio/exporters.py [hunk 2]
```diff
@@ -5,18 +5,25 @@
...
     from collections import OrderedDict
 except ImportError:
     # Python 2.6 compatibility
     from ordereddict import OrderedDict
 
+# Fix Python 2.x.
+try:
+    UNICODE_EXISTS = bool(type(unicode))
+except NameError:
+    unicode = lambda s: str(s)
```
Comment (davecap) [inline solvebio/exporters.py]: you can try using `six.u('asdfasdf')` instead of `unicode('asdfasdf')`

**Comment by jercoh** [comment]
@davecap pls review

File: solvebio/exporters.py [hunk 3]
```diff
@@ -247,6 +261,98 @@ def write(self, filename):
             f.close()
 
 
+class FlatCSVExporter(CSVExporter):
```
Comment (davecap) [inline solvebio/exporters.py]: do you think we should just replace the existing CSV and XLS exporters with this code?

File: solvebio/exporters.py [hunk 4]
```diff
@@ -247,6 +261,98 @@ def write(self, filename):
...
+        try:
+            return str(value)
+        except UnicodeEncodeError:
+            return ''.join(
+                [s.encode('ascii', 'backslashreplace') for s in value]
+            )
+
+    def write(self, filename):
+        if sys.version_info >= (3, 0, 0):
+            f = open(filename, 'w', newline='')
```
Comment (davecap) [inline solvebio/exporters.py]: is this newline='' correct? is it necessary?

File: solvebio/exporters.py [hunk 5]
```diff
@@ -247,6 +261,98 @@ def write(self, filename):
...
+        else:
+            f = open(filename, 'wb')
+
+        fieldnames = sorted(list(self.fields))
+        try:
+            writer = csv.DictWriter(f, fieldnames=fieldnames)
+            writer.writerow(dict(zip(fieldnames, fieldnames)))
```
Comment (davecap) [inline solvebio/exporters.py]: Is there a way to just get DictWriter to write the header? or is this the correct way?

File: solvebio/exporters.py [hunk 6]
```diff
@@ -247,6 +261,98 @@ def write(self, filename):
...
+            writer.writerows(self.rows)
+        except:
```
Comment (davecap) [inline solvebio/exporters.py]: you can just remove the `except: raise`

File: solvebio/test/test_exports.py [hunk 7]
```diff
@@ -20,6 +20,16 @@ def test_csv_exporter(self):
         self.assertTrue(path.isfile('/tmp/test.csv'))
         remove('/tmp/test.csv')
 
+    # FlatCSVExporter
+    def test_flat_csv_exporter(self):
+        dataset = Dataset.retrieve(self.TEST_DATASET_NAME)
+        query = dataset.query()[:10]
+
+        query.export('flat-csv', filename='/tmp/test_flat.csv')
+        self.assertTrue(path.isfile('/tmp/test_flat.csv'))
```
Comment (davecap) [inline solvebio/test/test_exports.py]: I guess we might want to test that the flat export is flattening correctly by comparing against a correct file, maybe by MD5 or by content.

File: solvebio/exporters.py (hunk 5 above)
Comment (jercoh) [inline solvebio/exporters.py]: I read this comment here: https://github.com/solvebio/solvebio-python/blob/28cf4e26a36bae2665dfd48ce0398e448546f8ee/solvebio/exporters.py#L252

File: solvebio/test/test_exports.py (hunk 7 above)
Comment (jercoh) [inline solvebio/test/test_exports.py]: yes we could do that for the other tests as well

**Pull request merged**